src/Shared/Microsoft.AspNetCore.Razor.Utilities.Shared/Utilities/CleanableWeakCache`2.cs

```diff
@@ -0,0 +1,297 @@
+    /// </summary>
+    /// <param name="key">The key of the value to get or add.</param>
+    /// <param name="valueFactory">A factory function to create the value if it needs to be added to the cache.</param>
+    /// <returns>
+    ///  The existing value if found and still alive, otherwise a new value created by <paramref name="valueFactory"/>.
+    /// </returns>
+    public TValue GetOrAdd(TKey key, Func<TValue> valueFactory)
+    {
+        // First check without creating the value.
+        lock (_lock)
+        {
+            if (TryGet_NoLock(key, out var value))
+            {
+                return value;
+            }
+        }
+
+        // Create the value outside the lock to avoid holding the lock
+        // while creating a potentially expensive object.
+        var newValue = valueFactory();
+
+        // Second check and add atomically
+        lock (_lock)
+        {
+            // Double-check in case another thread added it
+            if (TryGet_NoLock(key, out var existingValue))
+            {
+                return existingValue;
+            }
+
+            // Add our newly created value
+            TryAddOrGet_NoLock(key, newValue);
+            return newValue;
+        }
+    }
+
+    /// <summary>
+    ///  Gets the value associated with the specified key, or adds a new value created by the factory function
+    ///  using the provided argument if the key is not found or the previously cached value has been garbage collected.
+    /// </summary>
+    /// <typeparam name="TArg">The type of the argument passed to the value factory function.</typeparam>
+    /// <param name="key">The key of the value to get or add.</param>
+    /// <param name="arg">The argument to pass to the value factory function.</param>
+    /// <param name="valueFactory">A factory function to create the value using the provided argument.</param>
+    /// <returns>
+    ///  The existing value if found and still alive, otherwise a new value created by <paramref name="valueFactory"/>.
+    /// </returns>
+    public TValue GetOrAdd<TArg>(TKey key, TArg arg, Func<TArg, TValue> valueFactory)
+    {
+        // First check without creating the value.
+        lock (_lock)
+        {
+            // First, try to get an existing value
+            if (TryGet_NoLock(key, out var value))
+            {
+                return value;
+            }
+        }
+
+        // Create the value outside the lock to avoid holding the lock
+        // while creating a potentially expensive object.
+        var newValue = valueFactory(arg);
+
+        // Second check and add atomically
+        lock (_lock)
+        {
+            // Double-check in case another thread added it
+            if (TryGet_NoLock(key, out var existingValue))
+            {
+                return existingValue;
+            }
+
+            // Add our newly created value
+            TryAddOrGet_NoLock(key, newValue);
+            return newValue;
+        }
```

Comment:
Is the TryGet_NoLock call necessary? It feels like 

return TryAddOrGet_NoLock(key, newValue) ?? newValue;

would work. Or is the _addsSinceLastCleanUp increment the reason to avoid that? If that is the concern, would it make sense to move that increment/cleanup in TryAddOrGet_NoLock to the points where the value is added?